File: app/schemas/moderation.py
"""Pydantic schemas for reports and moderator quarantine workflows."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.quarantine_log import EntityType, ResolutionAction


class ReportCreate(BaseModel):
    """Optional reason when reporting a story part."""

    reason: str | None = Field(None, max_length=512)


class ReportResponse(BaseModel):
    """Confirmation after submitting a report."""

    id: UUID
    story_part_id: UUID
    reporter_id: UUID
    reason: str | None = None
    created_at: datetime
    quarantined: bool = False
    report_count: int = 1

    model_config = {"from_attributes": True}


class QuarantineLogResponse(BaseModel):
    """Quarantine log row for the moderator queue / audit log."""

    id: UUID
    entity_type: EntityType
    entity_id: UUID
    reason: str
    triggered_by: str
    automatic: bool
    resolved_by_moderator_id: UUID | None = None
    resolution_action: ResolutionAction | None = None
    resolved_at: datetime | None = None
    created_at: datetime
    # Enriched preview (populated for STORY_PART / USER when the entity still exists)
    author_username: str | None = None
    author_id: UUID | None = None
    teaser: str | None = None
    content_preview: str | None = None

    model_config = {"from_attributes": True}


class BlockUserRequest(BaseModel):
    """Optional reason when blocking a user."""

    reason: str | None = Field("Blocked by moderator", max_length=512)


class WarnUserRequest(BaseModel):
    """Warning message and optional temporary quarantine duration."""

    reason: str = Field(..., min_length=1, max_length=512)
    duration_hours: float | None = Field(None, gt=0, le=24 * 30)


class BulkModerationItem(BaseModel):
    """One target for a bulk moderation action."""

    entity_type: EntityType
    entity_id: UUID


class BulkModerationRequest(BaseModel):
    """Bulk allow / remove / block on queue items."""

    action: str = Field(..., pattern="^(allow|remove|block)$")
    items: list[BulkModerationItem] = Field(..., min_length=1, max_length=50)


class BulkModerationResponse(BaseModel):
    """Summary of a bulk moderation run."""

    processed: int
    failed: int
    errors: list[str] = []


class VotingPatternFlag(BaseModel):
    """A suspicious voting pattern for moderator review."""

    user_id: UUID
    username: str
    flag: str
    detail: str
    reputation_score: int


class ReputationPoint(BaseModel):
    """One point on a reputation history sparkline."""

    score: int
    created_at: datetime

File: frontend/src/api/moderation.ts
import apiClient from './client';

export interface QuarantineLog {
  id: string;
  entity_type: 'USER' | 'STORY_PART';
  entity_id: string;
  reason: string;
  triggered_by: string;
  automatic: boolean;
  resolved_by_moderator_id: string | null;
  resolution_action: 'ALLOWED' | 'REMOVED' | 'BLOCKED' | 'WARNED' | null;
  resolved_at: string | null;
  created_at: string;
  author_username?: string | null;
  author_id?: string | null;
  teaser?: string | null;
  content_preview?: string | null;
}

export interface ReportResponse {
  id: string;
  story_part_id: string;
  reporter_id: string;
  reason: string | null;
  created_at: string;
  quarantined: boolean;
  report_count: number;
}

export interface VotingPatternFlag {
  user_id: string;
  username: string;
  flag: string;
  detail: string;
  reputation_score: number;
  metric: number;
  severity: number;
}

export interface ReputationPoint {
  score: number;
  created_at: string;
}

export interface ModeratorUserVote {
  vote_id: string;
  vote_type: 'UP' | 'DOWN';
  voted_at: string;
  story_part_id: string;
  teaser: string;
  vote_score: number;
  recursive_score: number;
  is_quarantined: boolean;
  author_username: string | null;
}

export interface BulkModerationResult {
  processed: number;
  failed: number;
  errors: string[];
}

export interface ModeratorUserSummary {
  id: string;
  username: string;
  reputation_score: number;
  is_quarantined: boolean;
  is_blocked: boolean;
  is_moderator: boolean;
  created_at: string;
  last_activity_at: string;
}

export const moderatorApi = {
  getQueue: async (skip = 0, limit = 50): Promise<QuarantineLog[]> => {
    const response = await apiClient.get<QuarantineLog[]>('/moderator/quarantine-queue', {
      params: { skip, limit },
    });
    return response.data;
  },

  listUsers: async (skip = 0, limit = 50): Promise<ModeratorUserSummary[]> => {
    const response = await apiClient.get<ModeratorUserSummary[]>('/moderator/users', {
      params: { skip, limit },
    });
    return response.data;
  },

  getAuditLog: async (skip = 0, limit = 50): Promise<QuarantineLog[]> => {
    const response = await apiClient.get<QuarantineLog[]>('/moderator/audit-log', {
      params: { skip, limit },
    });
    return response.data;
  },

  getVotingPatterns: async (limit = 20): Promise<VotingPatternFlag[]> => {
    const response = await apiClient.get<VotingPatternFlag[]>('/moderator/voting-patterns', {
      params: { limit },
    });
    return response.data;
  },

  dismissVotingPattern: async (params: {
    userId: string;
    flag: string;
    reason?: string;
    /** Omit for server default; 0 = never expires. */
    durationHours?: number;
  }): Promise<VotingPatternFlag> => {
    const response = await apiClient.post<VotingPatternFlag>('/moderator/voting-patterns/dismiss', {
      user_id: params.userId,
      flag: params.flag,
      reason: params.reason || null,
      duration_hours: params.durationHours ?? null,
    });
    return response.data;
  },

  getReputationHistory: async (userId: string, limit = 50): Promise<ReputationPoint[]> => {
    const response = await apiClient.get<ReputationPoint[]>(
      `/moderator/users/${userId}/reputation-history`,
      { params: { limit } },
    );
    return response.data;
  },

  getUserVotes: async (
    userId: string,
    params: { skip?: number; limit?: number; vote_type?: 'UP' | 'DOWN' } = {},
  ): Promise<ModeratorUserVote[]> => {
    const response = await apiClient.get<ModeratorUserVote[]>(
      `/moderator/users/${userId}/votes`,
      {
        params: {
          skip: params.skip ?? 0,
          limit: params.limit ?? 20,
          vote_type: params.vote_type,
        },
      },
    );
    return response.data;
  },

  bulk: async (
    action: 'allow' | 'remove' | 'block',
    items: { entity_type: string; entity_id: string }[],
  ): Promise<BulkModerationResult> => {
    const response = await apiClient.post<BulkModerationResult>('/moderator/bulk', {
      action,
      items,
    });
    return response.data;
  },

  quarantineStory: async (storyId: string, reason?: string): Promise<QuarantineLog> => {
    const response = await apiClient.post<QuarantineLog>(
      `/moderator/stories/${storyId}/quarantine`,
      { reason: reason ?? null },
    );
    return response.data;
  },

  allow: async (entityType: string, entityId: string): Promise<QuarantineLog> => {
    const response = await apiClient.post<QuarantineLog>(
      `/moderator/${entityType}/${entityId}/allow`,
    );
    return response.data;
  },

  remove: async (entityType: string, entityId: string): Promise<QuarantineLog> => {
    const response = await apiClient.post<QuarantineLog>(
      `/moderator/${entityType}/${entityId}/remove`,
    );
    return response.data;
  },

  blockUser: async (userId: string, reason?: string): Promise<QuarantineLog> => {
    const response = await apiClient.post<QuarantineLog>(`/moderator/users/${userId}/block`, {
      reason: reason ?? 'Blocked by moderator',
    });
    return response.data;
  },

  unblockUser: async (userId: string, reason?: string): Promise<QuarantineLog> => {
    const response = await apiClient.post<QuarantineLog>(`/moderator/users/${userId}/unblock`, {
      reason: reason ?? 'Unblocked by moderator',
    });
    return response.data;
  },

  warnUser: async (
    userId: string,
    reason: string,
    durationHours?: number,
  ): Promise<QuarantineLog> => {
    const response = await apiClient.post<QuarantineLog>(`/moderator/users/${userId}/warn`, {
      reason,
      duration_hours: durationHours ?? null,
    });
    return response.data;
  },
};

export const reportStoryPart = async (
  storyId: string,
  reason?: string,
): Promise<ReportResponse> => {
  const response = await apiClient.post<ReportResponse>(`/stories/${storyId}/report`, {
    reason: reason ?? null,
  });
  return response.data;
};
